Sources/PackageLoading/RegistryReleaseMetadataSerialization.swift

```diff
@@ -0,0 +1,173 @@
+        let codableMetadata = CodableRegistryReleaseMetadata(metadata)
+        let data = try Self.encoder.encode(codableMetadata)
+        try fileSystem.writeFileContents(path, data: data)
+    }
+
+    public static func load(from path: AbsolutePath, fileSystem: FileSystem) throws -> RegistryReleaseMetadata {
+        let codableMetadata = try Self.decoder.decode(
+            path: path,
+            fileSystem: fileSystem,
+            as: CodableRegistryReleaseMetadata.self
+        )
+        return try RegistryReleaseMetadata(codableMetadata)
+    }
+}
+
+private struct CodableRegistryReleaseMetadata: Codable {
+    public let registry: URL
+    public let signature: Signature?
+    public let author: Author?
+    public let description: String?
+    public let licenseURL: URL?
+    public let readmeURL: URL?
+    public let scmRepositoryURLs: [URL]?
+
+    init(_ seed: RegistryReleaseMetadata) {
+        switch seed.source {
+        case .registry(let url):
+            self.registry = url
+        }
+        self.signature = seed.signature.map { signature in
+            .init(
+                signedBy: signature.signedBy.flatMap {
+                    switch $0 {
+                    case .recognized(let type, let commonName, let organization, let identity):
+                        return .recognized(
+                            type: type,
+                            commonName: commonName,
+                            organization: organization,
+                            identity: identity
+                        )
+                    case .unrecognized(let commonName, let organization):
+                        return .unrecognized(commonName: commonName, organization: organization)
+                    }
+                },
+                format: signature.format,
+                base64: Data(signature.value).base64EncodedString()
+            )
+        }
+        self.author = seed.metadata.author.map {
+            .init(
+                name: $0.name,
+                emailAddress: $0.emailAddress,
+                description: $0.description,
+                url: $0.url,
+                organization: $0.organization.map {
+                    .init(
+                        name: $0.name,
+                        emailAddress: $0.emailAddress,
+                        description: $0.description,
+                        url: $0.url
+                    )
+                }
+            )
+        }
+        self.description = seed.metadata.description
+        self.licenseURL = seed.metadata.licenseURL
+        self.readmeURL = seed.metadata.readmeURL
+        self.scmRepositoryURLs = seed.metadata.scmRepositoryURLs
+    }
+
+    public struct Signature: Codable {
+        let signedBy: SigningEntity?
+        let format: String
+        let base64: String
+    }
+
+    public enum SigningEntity: Codable {
+        case recognized(type: String?, commonName: String?, organization: String?, identity: String?)
```

Comment:
`type` can be non-optional I think

Reply:
these are all optional on PackageSigning.SigningEntity causing this to also need to be optional. I guess we could also change PackageSigning.SigningEntity to carry these as associated values which will help. is this what you have in mind?

Reply:
addressed for now, perhaps we can make PackageSigning.SigningEntity an enum with associated values, I think it may model things nicely